fuel/datasets/hdf5.py

```diff
@@ -316,11 +316,14 @@ def _out_of_memory_get_data(self, state=None, request=None):
             if isinstance(request, slice):
                 request = slice(request.start + subset.start,
                                 request.stop + subset.start, request.step)
+                data = state[source_name][request]
             elif isinstance(request, list):
                 request = [index + subset.start for index in request]
+                sorted_request = numpy.sort(request)
+                sorted_data = state[source_name][sorted_request, ...]
+                data = sorted_data[numpy.argsort(numpy.argsort(request))]
```

Comment:
Doing 3 sorts seems a bit crazy. I guess you could two:
```python
indices = numpy.argsort(request)
reverse_indices = numpy.argsort(indices)
sorted_data = state[source_name][request[indices], ...]
data = sorted_data[reverse_indices]
```
Maybe there's a way of doing it with only 1 as well?